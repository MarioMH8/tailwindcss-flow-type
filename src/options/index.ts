export { default as DEFAULT_OPTIONS } from './options.default';
export { default as parseOptions } from './options.functions';
export * from './options.type';
