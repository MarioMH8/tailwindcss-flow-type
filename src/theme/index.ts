export { default as createTextTheme } from './create-text-theme';
export { FLOW_TYPE_THEME_LINE_HEIGHT_DEFAULT, FLOW_TYPE_THEME_TEXT_DEFAULT } from './flow-theme.default';
export type { TailwindFlowTypeTheme } from './flow-theme.type';
export { TAILWIND_FLOW_TYPE_SIZES } from './flow-theme.type';
