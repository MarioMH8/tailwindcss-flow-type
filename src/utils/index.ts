export { default as calculateModularScale } from './calculate-modular-scale';
export { default as isNumber } from './is-number';
