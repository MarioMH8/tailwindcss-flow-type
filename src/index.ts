import type { ThemeConfig } from 'tailwindcss/plugin';
import createPlugin from 'tailwindcss/plugin';

import type { TailwindFlowTypeOptions } from './options';
import { parseOptions } from './options';
import { createTextTheme, FLOW_TYPE_THEME_LINE_HEIGHT_DEFAULT, FLOW_TYPE_THEME_TEXT_DEFAULT } from './theme';
import { calculateModularScale, isNumber } from './utils';

export * from './options';
export * from './theme';
export * from './utils';

export default createPlugin.withOptions<Partial<TailwindFlowTypeOptions>>(
	userOptions => {
		const options = parseOptions(userOptions);

		return tailwind => {
			tailwind.matchUtilities(
				{
					[`${options.prefix}-text`]: value => {
						const fontSize =
							typeof value === 'string' ? (isNumber(value) ? Number(value) : value) : undefined;

						if (fontSize === undefined) {
							return {};
						}

						return {
							fontSize:
								typeof fontSize === 'number' ? calculateModularScale(fontSize, options) : fontSize,
						};
					},
				},
				{
					supportsNegativeValues: false,
					values: tailwind.theme('flow-text') as Record<string, string>,
				}
			);
			tailwind.matchUtilities(
				{
					[`${options.prefix}-text`]: value => {
						return {
							lineHeight: value,
						};
					},
				},
				{
					supportsNegativeValues: false,
					values: tailwind.theme('flow-line-height') as Record<string, string>,
				}
			);
		};
	},
	userOptions => {
		const options = parseOptions(userOptions);

		const theme: ThemeConfig = {
			'flow-line-height': FLOW_TYPE_THEME_LINE_HEIGHT_DEFAULT,
			'flow-text': FLOW_TYPE_THEME_TEXT_DEFAULT,
		};

		if (options.override) {
			theme['text'] = createTextTheme(options);
		}

		return {
			theme,
		};
	}
);
