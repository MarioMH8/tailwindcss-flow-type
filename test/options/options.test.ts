import { describe, expect, it } from 'bun:test';

import type { TailwindFlowTypeOptions } from '../../src';
import { DEFAULT_OPTIONS, parseOptions, TAILWIND_FLOW_TYPE_SIZES } from '../../src';

describe('tailwindcss-flow-type', () => {
	describe('options', () => {
		describe('DEFAULT_OPTIONS', () => {
			it('not change', () => {
				expect(DEFAULT_OPTIONS).toMatchSnapshot();
			});
		});
		describe('TAILWIND_FLOW_TYPES_SIZES', () => {
			it('not change', () => {
				expect(TAILWIND_FLOW_TYPE_SIZES).toMatchSnapshot();
			});
		});
		describe('parseOptions', () => {
			it('should works as expected when no user options are provided', () => {
				const result = parseOptions();
				expect(result).toStrictEqual(DEFAULT_OPTIONS);
			});

			it('should works as expected when provide settings', () => {
				const options: TailwindFlowTypeOptions = {
					fontSizeMax: 3,
					fontSizeMin: 1.5,
					override: true,
					prefix: 'ggg',
					ratioMax: 1.8,
					ratioMin: 1.2,
					screenMax: 1440,
					screenMin: 360,
					unit: 'em',
				};
				const result = parseOptions(options);
				expect(result.fontSizeMax).toBe(3);
				expect(result.fontSizeMin).toBe(1.5);
				expect(result.ratioMax).toBe(1.8);
				expect(result.ratioMin).toBe(1.2);
				expect(result.screenMax).toBe(1440);
				expect(result.screenMin).toBe(360);
				expect(result.override).toBe(true);
				expect(result.prefix).toBe('ggg');
				expect(result.unit).toBe('em');
				expect(result).toMatchSnapshot();
			});
		});
	});
});
