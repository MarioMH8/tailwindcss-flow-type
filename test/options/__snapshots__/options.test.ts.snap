// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`tailwindcss-flow-type options DEFAULT_OPTIONS not change 1`] = `
{
  "fontSizeMax": 1.25,
  "fontSizeMin": 1.125,
  "override": false,
  "prefix": "flow",
  "ratioMax": 1.2,
  "ratioMin": 1.125,
  "screenMax": 96,
  "screenMin": 20,
  "unit": "rem",
}
`;

exports[`tailwindcss-flow-type options TAILWIND_FLOW_TYPES_SIZES not change 1`] = `
[
  "2xl",
  "3xl",
  "4xl",
  "5xl",
  "6xl",
  "7xl",
  "8xl",
  "9xl",
  "base",
  "lg",
  "sm",
  "xl",
  "xs",
]
`;

exports[`tailwindcss-flow-type options parseOptions should works as expected when provide settings 1`] = `
{
  "fontSizeMax": 3,
  "fontSizeMin": 1.5,
  "override": true,
  "prefix": "ggg",
  "ratioMax": 1.8,
  "ratioMin": 1.2,
  "screenMax": 1440,
  "screenMin": 360,
  "unit": "em",
}
`;
